heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,623 @@
+typedef struct list64_t {
+  int64_t data;
+  struct list64_t *next;
+} list64_t;
+
+int64_t is_elem (int64_t x, list64_t *l) {
+  if (l == NULL) {
+    return 0;
+  } else if (l->data == x) {
+    return 1;
+  } else {
+    return is_elem (x, l->next);
+  }
+}
+```
+
+#### 1. Generating LLVM bitcode
+
+Just like with SAW, we want to work with the LLVM bitcode (`.bc`), so
+you can generate it just like before. Note that we have already
+included the binary for all the examples so you don't really need to
+run this command.
+
+```bash
+clang -g -c -emit-llvm -o examples/linked_list.bc examples/linked_list.c
+```
+
+(The alternative command `make examples/linked_list.bc` won't work
+because the shorcut is not defined in the Makefile of heapster.)
+
+Be aware that the resulting bitcode may depend on your `clang` version and your
+operating system. In turn, this means the Heapster commands in your SAW script
+and the proofs in your Coq file may also be dependent on how and where the
+bitcode is generated. If you find an incompatibility, please report it.
+
+#### 2. Run the SAW interpreter with Heapster
+
+We start by running saw with `cabal run saw`. Once SAW is loaded, you
+can load all the Heapster commands with
+
+```bash
+sawscript> enable_experimental
+```
+
+If you print the environment now (wiht `:env`) you will notice a new
+set of commands, all starting with `heapster_*`. You can also start
+typing the name and press tab to see all the functions those are all the Heapster commands.
+
+```
+sawscript> heapster_ [TAB]
+heapster_assume_fun                  heapster_find_symbols
+heapster_assume_fun_multi            heapster_find_symbols_with_type
+heapster_assume_fun_rename           heapster_find_trait_method_symbol
+heapster_assume_fun_rename_prim      heapster_gen_block_perms_hint
+heapster_block_entry_hint            heapster_get_cfg
+heapster_define_irt_recursive_perm   heapster_init_env
+heapster_define_irt_recursive_shape  heapster_init_env_debug
+heapster_define_llvmshape            heapster_init_env_for_files
+heapster_define_opaque_llvmshape     heapster_init_env_for_files_debug
+heapster_define_opaque_perm          heapster_init_env_from_file
+heapster_define_perm                 heapster_init_env_from_file_debug
+heapster_define_reachability_perm    heapster_join_point_hint
+heapster_define_recursive_perm       heapster_parse_test
+heapster_define_rust_type            heapster_print_fun_trans
+heapster_define_rust_type_qual       heapster_set_debug_level
+heapster_export_coq                  heapster_set_translation_checks
+heapster_find_symbol                 heapster_typecheck_fun
+heapster_find_symbol_commands        heapster_typecheck_fun_rename
+heapster_find_symbol_with_type       heapster_typecheck_mut_funs
+```
+
+You can then use `:?` to see further information for each of them.
+
+
+#### 3. Load the file and extract the two function specifications.
+
+To load a file into heapster you can use `heapster_init_env`. Let's
```

Comment:
You should maybe introduce `heapster_init_env_from_file` first, because almost all of the examples (including specifically the `linked_list` example) use it, and then you can add an extra few sentences mentioning that `heapster_init_env` is a special case